lib/FrontendTool/FrontendTool.cpp

```diff
@@ -1290,10 +1290,16 @@ static bool performCompile(CompilerInstance &Instance,
   if (opts.PrintClangStats && Context.getClangModuleLoader())
     Context.getClangModuleLoader()->printStatistics();
 
-  emitReferenceDependenciesForAllPrimaryInputsIfNeeded(Instance);
   emitSwiftRangesForAllPrimaryInputsIfNeeded(Instance);
   emitCompiledSourceForAllPrimaryInputsIfNeeded(Instance);
 
+  SWIFT_DEFER {
+    // We might have freed the ASTContext already, but in that case we must have
+    // emitted the dependencies first.
+    if (Instance.hasASTContext())
+      emitReferenceDependenciesForAllPrimaryInputsIfNeeded(Instance);
```

Comment:
I think we should work to remove the dependency on ASTContext from this job. It's really only needed because the referenced name trackers live there.

Reply:
We need the ASTContext and the Decls it owns when writing out dependencies because we mangle names, etc. I agree in principle we should untangle this, but it's a larger refactoring.

Reply:
(Ideally, we would write out the 'provides' set after type checking and the 'depends' set after IRGen, as we discussed in the past).